protocols/raft/src/main/java/io/atomix/protocols/raft/storage/snapshot/impl/DefaultSnapshot.java

```diff
@@ -0,0 +1,198 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License
+ */
+package io.atomix.protocols.raft.storage.snapshot.impl;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import io.atomix.protocols.raft.storage.snapshot.Snapshot;
+import io.atomix.protocols.raft.storage.snapshot.SnapshotReader;
+import io.atomix.protocols.raft.storage.snapshot.SnapshotWriter;
+import io.atomix.protocols.raft.storage.snapshot.Snapshottable;
+import io.atomix.utils.time.WallClockTimestamp;
+import java.util.Objects;
+
+/**
+ * Manages reading and writing a single snapshot file.
+ *
+ * <p>User-provided state machines which implement the {@link Snapshottable} interface transparently
+ * write snapshots to and read snapshots from files on disk. Each time a snapshot is taken of the
+ * state machine state, the snapshot will be written to a single file represented by this interface.
+ * Snapshots are backed by a {@link io.atomix.storage.buffer.Buffer} dictated by the parent {@link
+ * io.atomix.storage.StorageLevel} configuration. Snapshots for file-based storage levels like
+ * {@link io.atomix.storage.StorageLevel#DISK DISK} will be stored in a disk backed buffer, and
+ * {@link io.atomix.storage.StorageLevel#MEMORY MEMORY} snapshots will be stored in an on-heap
+ * buffer.
+ *
+ * <p>Snapshots are read and written by a {@link SnapshotReader} and {@link SnapshotWriter}
+ * respectively. To create a reader or writer, use the {@link #openReader()} and {@link
+ * #openWriter()} methods.
+ *
+ * <p>
+ *
+ * <pre>{@code
+ * Snapshot snapshot = snapshotStore.snapshot(1);
+ * try (SnapshotWriter writer = snapshot.writer()) {
+ *   writer.writeString("Hello world!");
+ * }
+ * snapshot.complete();
+ *
+ * }</pre>
+ *
+ * A {@link SnapshotReader} is not allowed to be created until a {@link SnapshotWriter} has
+ * completed writing the snapshot file and the snapshot has been marked {@link #complete()
+ * complete}. This allows snapshots to effectively be written and closed but not completed until
+ * other conditions are met. Prior to the completion of a snapshot, a failure and recovery of the
+ * parent {@link DefaultSnapshotStore} will <em>not</em> recover an incomplete snapshot. Once a
+ * snapshot is complete, the snapshot becomes immutable, can be recovered after a failure, and can
+ * be read by multiple readers concurrently.
+ */
+public abstract class DefaultSnapshot implements Snapshot {
```

Comment:
I'm somehow allergic against these `Default*` naming. Could we maybe just rename it to FileSnapshot or SingleFileSnapshot idk. :man_shrugging:

Reply:
:sweat_smile: I just followed the Atomix "convention". `FileSnapshot` already exists as well hehe

I guess the `Default*` is used to denote the basic implementation, other things I've seen are `*Impl` versions for the default ones. I have no strong preference tbh